HP-318 step 2: GraphQL update

NikoHelle opened the pull request:
Apollo client was old and deprecated and did not support cache updating. Changed to client to new one - from same authors so switch was easy.

Updated MyProfileQuery to get Verified Personal Information and render it with a new component. Added suomi.fi as type of authorisation where needed.

NikoHelle committed:
Upadated apollo client and removed apollo-boost

Old client is depricated and does not support cache modifications. Apollo boost is not needed anymore

NikoHelle committed:
Include Verified personal information in GraphQLQuery.

Ignore error received when using non-verified IDP.

GraphQL data structure has been changed, so it was needed to run codegen that updates generatedTypes. Codegen does not work with current versions, because node_modules have multiple versions of GraphQL package. "Resolutions" in package.json resolves that with some package updates.

NikoHelle committed:
Render verified PI if received from BE.

Updated translation files.

NikoHelle committed:
Show suomi.fi as an AMR.

Also DownloadData and DeleteProfile components inside ProfileInformation cause unnecessary graphQL reloads when Profile changes. Moved their code to independent components to avoid reloading

NikoHelle committed:
Small css and translation update

Missing translation added
Small css fixes..

akikoskinen commented:
@NikoHelle can you rebase on top of the current `develop` branch?

NikoHelle commented:
> @NikoHelle can you rebase on top of the current `develop` branch?

Done!

akikoskinen commented on yarn.lock:
```diff
@@ -3424,6 +3452,11 @@ autoprefixer@^9.6.1:
+  version "1.1.2"
+  resolved "https://registry.yarnpkg.com/await-handler/-/await-handler-1.1.2.tgz#329c19e9382599898bc12c80c47efdcdaa1b4992"
+  integrity sha512-dihteGhwbJpT89kVbacWiyKeAZr+En0YGK6pAKQJLR0En9ZxSH2H4TTvfG4bBjzFq9gDAma4y9BrpDns6j5UiQ==
+
```
`await-handler` dependency is added in commit 475885465bb...

akikoskinen commented on yarn.lock:
```diff
@@ -3452,11 +3452,6 @@ autoprefixer@^9.6.1:
-  version "1.1.2"
-  resolved "https://registry.yarnpkg.com/await-handler/-/await-handler-1.1.2.tgz#329c19e9382599898bc12c80c47efdcdaa1b4992"
-  integrity sha512-dihteGhwbJpT89kVbacWiyKeAZr+En0YGK6pAKQJLR0En9ZxSH2H4TTvfG4bBjzFq9gDAma4y9BrpDns6j5UiQ==
-
```
...`await-handler` dependency is removed in commit f8f631bc8a80. And that's the only contents of this commit. Can commit 475885465bb be changed so that `await-handler` dependency is not added (apparently it's not needed)? And commit f8f631bc8a80 can thus be removed.

NikoHelle replied:
Later on, in next PR, await-hander is needed and will be added. I removed it from package.json for this PR, but forgot to run "yarn" to remove it from yarn.lock. But then tests fail, because yarn.lock is not in sync and "yarn install" in test container is trying to create new yarn.lock without write privileges. So I had to update yarn.lock.

NikoHelle replied:
f8f6 is now removed. await-handler was removed in a new commit that updated yarn.lock anyway and await-handler in not seen anymore as a change.

akikoskinen replied:
`await-handler` is still added to, and then removed from `yarn.lock`, in this PR. But `await-handler` is not in `package.json` in any commit in this PR. I don't understand why it has to be in `yarn.lock` for a few commits. What brings it there?

NikoHelle replied:
yarn.lock is an autogenerated file and should not be edited by hand. await-handler was previously in package.json, so it is also here. This commit was edited afterwards. But lock file do not change unless yarn in also run. Running yarn separately just for this commit is pointless.

NikoHelle replied:
All yarn.lock updates are now reset from commits and created a new commit just for new yarn.lock.

akikoskinen replied:
> yarn.lock is an autogenerated file and should not be edited by hand.

Yes, that's clear.

> await-handler was previously in package.json, so it is also here.

What do you mean by "was previously"? Running `git log -S await-handler` while this PR's branch is checked out, gives no results, so it's not found anywhere in the commit history.

> This commit was edited afterwards. But lock file do not change unless yarn in also run. Running yarn separately just for this commit is pointless.

No. Yarn (or whatever package management tool is used) **must** be run every time the dependency list file (e.g. `package.json`) is changed. And the lock file **must** be included in the same commit. If yarn was not run for a commit where `package.json` was changed, you use interactive rebase to go back to that commit that modified `package.json`, run yarn and rewrite the commit adding the changed lock file.

> All yarn.lock updates are now reset from commits and created a new commit just for new yarn.lock.

That's wrong. `Package.json` and related `yarn.lock` changes need to be in the same commit. You can rewrite history as long as this is a PR we are talking about.

NikoHelle replied:
await-handler cannot be found in commit history, because that change in package.json is not in any commit anymore. That caused this discussion as it was in one commit and removed in another. 

yarn.lock in now back in each commit that alters package.json.

akikoskinen commented on .env.development:
```diff
@@ -1,3 +1,5 @@
-REACT_APP_OIDC_AUTHORITY="https://api.hel.fi/sso-test/"
-REACT_APP_OIDC_CLIENT_ID="https://api.hel.fi/auth/helsinkiprofile-ui"
-REACT_APP_PROFILE_GRAPHQL="https://profiili-api.test.kuva.hel.ninja/graphql/"
+REACT_APP_OIDC_AUTHORITY="https://tunnistamo.dev.hel.ninja/"
+REACT_APP_OIDC_CLIENT_ID="https://api.hel.fi/auth/helsinkiprofile-ui-dev"
+REACT_APP_PROFILE_AUDIENCE="https://api.hel.fi/auth/helsinkiprofiledev"
+REACT_APP_PROFILE_GRAPHQL="https://helsinki-profile-api-dev.agw.arodevtest.hel.fi/graphql/"
+REACT_APP_OIDC_SCOPE="openid profile https://api.hel.fi/auth/helsinkiprofiledev"
```
Is this (and the corresponding `public/env-config.js` change) somehow related to the apollo client update, as these changes are now in the same commit? I don't understand the relevance.

NikoHelle replied:
Previous, very old, configs did not provide a login flow that results in getting Verified Personal Information. Time and need to update them to use OpenShift.

akikoskinen replied:
But could the config be changed in its own commit, if it's not tightly related to the apollo client update?

NikoHelle replied:
Sliced and diced the file (+ auto-generated env-config.js) to another commit.

akikoskinen commented on src/graphql/generatedTypes.ts:
```diff
@@ -298,8 +301,100 @@ export interface MyProfileQuery_myProfile_phones {
+
 export interface MyProfileQuery_myProfile {
-  readonly __typename: "ProfileNode";
+  readonly __typename: "ProfileWithVerifiedPersonalInformationNode";
```
In a [very recent version of open-city-profile](https://github.com/City-of-Helsinki/open-city-profile/commit/9fdfb228de57117a362ff8051ae18097197ca718), the type returned by `myProfile` query is once again called `ProfileNode`. Perhaps calling it temporarily `ProfileWithVerifiedPersonalInformationNode` is not needed in this UI code.

NikoHelle replied:
This is an autogenerated file from a script that gets its data from graphQL-server. Script hasn't been run for a while. I'll run it again and reflect changes to the codebase.

NikoHelle replied:
Done.

akikoskinen replied:
Could squash the fixing commit to the one that introduces the `ProfileWithVerifiedPersonalInformationNode` names in the first place.

NikoHelle replied:
This has been done yesterday.

akikoskinen commented on src/i18n/en.json:
```diff
@@ -135,16 +161,37 @@
     "created": "Created:",
     "empty": "You currently don't have any services connected to your profile.",
-    "explanation": "EN: Olet antanut seuraaville palveluille oikeuden käyttää Helsinki Profiili -tietojasi.",
+    "explanation": "You have given the following services permission to use you Helsinki Profile data.",
```
Typo: second "you" should probably be "your".

NikoHelle replied:
Done. This is update is from somewhere else... But all translation files are updated simultaneously.

akikoskinen commented on src/i18n/en.json:
```diff
@@ -135,16 +161,37 @@
+    "verifiedDataInformationPrefix": "EN: Merkityt",
+    "verifiedDataInformationAriaPrefix": "EN: Alla olevat",
+    "verifiedDataInformation": "EN: <0></0> tiedot on vahvistettu virallisesta väestötietojärjestelmästä, eikä niitä ei voi tämän takia muokata Helsinki Profiilissa. Voit tutustua halutessasi väestötietojärjestelmään tallennettuihin tietoihisi suomi.fi verkkopalvelussa: <1>siirry suomi.fi palveluun </1>",
+    "verifiedDataInformationLink": "https://testipalvelu.apro.tunnistus.fi",
```
I don't know whether this URL needs to be localizable. But at least it needs to be configurable in different deployment environments. I have no idea how to best achieve that, though.

NikoHelle replied:
Yes, there is no way to make environment dependant translations at the moment. It should be done when there is actual need. Maybe all links can just point to production versions until other versions are actually required. This url was given to me, because suomi.fi was apparently wrong.

akikoskinen commented on src/profile/components/verifiedPersonalInformation/VerifiedPersonalInformation.tsx:
```diff
@@ -0,0 +1,197 @@
+    <ProfileSection hasVerifiedUserData>
+      <h3
+        className={commonFormStyles.sectionTitle}
+        aria-label="Vahvistetut perustiedot"
```
Should `aria-label`s be localized?

akikoskinen replied:
Aha, it **is** localized, but in another commit.

akikoskinen commented on src/profile/helpers/getCountry.ts:
```diff
@@ -0,0 +1,11 @@
+import getLanguageCode from '../../common/helpers/getLanguageCode';
+import { MyProfileQuery_myProfile_addresses_edges_node as AddressNode } from '../../graphql/generatedTypes';
+
+export default function getCountry(
+  countryCode: AddressNode['countryCode'] | undefined,
+  lang: string
+): string {
+  return countries.getName(countryCode || 'FI', getLanguageCode(lang));
+}
```
Ooooh, internationalized country names. Very nice 👍

akikoskinen commented on src/config.ts:
```diff
@@ -32,6 +32,10 @@ const config = {
     'https://tunnus.yle.fi/#omat-tiedot'
   ),
+  identityProviderManagementUrlTunnistusSuomifi: defaultTo(
+    window._env_.REACT_APP_IPD_MANAGEMENT_URL_TUNNISTUS_SUOMI_FI,
```
It seems these configurations have been mentioned in the README. Perhaps this new one should be added there too.

NikoHelle replied:
Done.

akikoskinen commented on src/profile/components/deleteProfile/DeleteProfile.tsx:
```diff
@@ -4,34 +4,50 @@ import { loader } from 'graphql.macro';
 import useToast from '../../../toast/useToast';
 import styles from './deleteProfile.module.css';
+import useDeleteProfile from '../../../gdprApi/useDeleteProfile';
+import checkBerthError from '../../helpers/checkBerthError';
+import BerthErrorModal from '../modals/berthError/BerthErrorModal';
```
I wonder what all these \*berth\* things are... But I see things were just moved around, so no need to do anything about that in this PR. But perhaps need to write it down somewhere that there could be something that should be checked if it's still correct (as there are and will be other services than just berth).

NikoHelle replied:
Good point!

akikoskinen reviewed (changes requested):
I would expect there to be some tests as there is new functionality.

NikoHelle committed:
Added new configuration variable documentation into README

NikoHelle committed:
Updated development configs to use OpenShift servers

NikoHelle commented:
> I would expect there to be some tests as there is new functionality.

Tests are done, but added in the next PR because they depend on libraries and data mocking that are not part of Apollo update.

akikoskinen commented:
> > I would expect there to be some tests as there is new functionality.
> 
> Tests are done, but added in the next PR because they depend on libraries and data mocking that are not part of Apollo update.

That makes it a bit hard to do a proper review as tests and production code always go hand in hand. Now I can't see the other hand.

NikoHelle commented:
> > > I would expect there to be some tests as there is new functionality.
> > 
> > 
> > Tests are done, but added in the next PR because they depend on libraries and data mocking that are not part of Apollo update.
> 
> That makes it a bit hard to do a proper review as tests and production code always go hand in hand. Now I can't see the other hand.

New functionalities are in components that only render text and 2 links: VerifiedPersonalInformation and  ProfileInformationAccountManagementLink. No new logic. That's why tests are not included here. I can copy tests from next PR, but it requires a lot of work getting them isolated from other tests and exclude the code not needed in this PR.

NikoHelle commented:
> > > > I would expect there to be some tests as there is new functionality.
> > > 
> > > 
> > > Tests are done, but added in the next PR because they depend on libraries and data mocking that are not part of Apollo update.
> > 
> > 
> > That makes it a bit hard to do a proper review as tests and production code always go hand in hand. Now I can't see the other hand.
> 
> New functionalities are in components that only render text and 2 links: VerifiedPersonalInformation and ProfileInformationAccountManagementLink. No new logic. That's why tests are not included here. I can copy tests from next PR, but it requires a lot of work getting them isolated from other tests and exclude the code not needed in this PR.

NikoHelle committed:
Test update

Added tests for VerifiedProfileInformation. Requires mocked personal data, done in myProfileMocking.ts. Added new testing library for easier element queries from DOM.

Added suomi.fi test to ProfileInformationAccountManagementLinks.
